flutter/flutter-provider/state-management-in-flutter-using-provider-Part2.md

```diff
@@ -0,0 +1,198 @@
+2. What are the types of Provider?
+3. How to use Provider in our code?
+4. What is Consumer & Selector?
+  
+### 1. What is Provider, Why we use it
+
+`Provider` is one of the most popular state management solution, created by  [Remi Rousselet](https://github.com/rrousselGit/provider), can be used in flutter app development.
+
+On the beginner stage, we use `setState` to update the app's state, because it is easy, and if we try to use some other moderate or say a complex technique like `flutter_bloc` or `Mobx`, this is not easy to switch from `setState` to these state management technique 😓
+
+Provider uses InheritedWidget internally, to manage all complex computations and provide a simple feature to manage our app's state
+
+>Why `provider`? We have some other techniques that can be used for managing state.
+
+Passing data via constructors is not a good idea, in doing so, all those widgets will rebuild just for passing the data. This may cause the performance issue and our code get too much complicated.
+
+So, what we should do now? We will use the provider
+
+>`flutter_bloc` & `Mobx` uses provider internally 🥳 And now the flutter team also supports to use provider.
+
+So, let's learn about the types of provider
+
+### 2. What are the types of provider
+
+Provider can handle different type of objects
+
+Complete [list of all objects](https://pub.dev/documentation/provider/latest/provider/provider-library.html)
+
+| Name|Description|
+|-----|-----------|
+|Provider|The most basic form of provider. It takes a value and exposes it, whatever the value is.|
+|ListenableProvider|A specific provider for Listenable object. ListenableProvider will listen to the object and ask widgets which depend on it to rebuild whenever the listener is called.|
+|ChangeNotifierProvider|A specification of ListenableProvider for ChangeNotifier. It will automatically call `ChangeNotifier.dispose` when needed.|
+|ValueListenableProvider|Listen to a ValueListenable and only expose `ValueListenable.value`.|
+|StreamProvider|Listen to a Stream and expose the latest value emitted.|
+|FutureProvider|Takes a `Future` and updates dependents when the future completes.|
+
+So, let's discover more about the categories
+
+### Provider
+
+This is a very basic form of `Provider`, manages the lifecycle of the value it provides by delegating to a pair of `Create` and `Dispose`. This can provide any type of value.
+
+```dart
+class Counter {
+    int count = 0;
+
+  void increment() {
+      count += 1;
+      notifyListeners();
+  }
+}
+
+class Stateless extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return Provider<Counter>(
+      create: (context) =>  Counter(),
+      dispose: (context, value) => value.dispose(),
+      child: ...,
+    );
+  }
+}
+
+```
+
+### ListenableProvider
+
+ListenableProvider is a bit different from the other providers in the list, as this provider provides `state source` as the value.
```

Comment:
Add backticks to code keywords. ListenableProvider should be `ListenableProvider`